File: src/com/crossbowffs/quotelock/modules/freakuotes/FreakuotesQuoteModule.java
package com.crossbowffs.quotelock.modules.freakuotes;

import android.content.ComponentName;
import android.content.Context;
import android.text.TextUtils;
import com.crossbowffs.quotelock.R;
import com.crossbowffs.quotelock.api.QuoteData;
import com.crossbowffs.quotelock.api.QuoteModule;
import com.crossbowffs.quotelock.utils.IOUtils;
import com.crossbowffs.quotelock.utils.Xlog;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

public class FreakuotesQuoteModule implements QuoteModule {
    private static final String TAG = FreakuotesQuoteModule.class.getSimpleName();

    @Override
    public String getDisplayName(Context context) {
        return context.getString(R.string.module_freakuotes_name);
    }

    @Override
    public ComponentName getConfigActivity(Context context) {
        return null;
    }

    @Override
    public int getMinimumRefreshInterval(Context context) {
        return 0;
    }

    @Override
    public boolean requiresInternetConnectivity(Context context) {
        return true;
    }

    @Override
    public QuoteData getQuote(Context context) throws Exception {
        String html = IOUtils.downloadString("https://freakuotes.com/frase/aleatoria");
        Document document = Jsoup.parse(html);
        Element quoteContainer = document.select(".quote-container > blockquote").first();

        String quoteText = quoteContainer.getElementsByTag("p").text();
        if (TextUtils.isEmpty(quoteText)) {
            Xlog.e(TAG, "Failed to find quote text");
            return null;
        }

        String sourceLeft = quoteContainer.select("footer > span").text();
        String sourceRight = quoteContainer.select("footer > cite").attr("title");
        String quoteSource;
        if (TextUtils.isEmpty(sourceLeft) && TextUtils.isEmpty(sourceRight)) {
            Xlog.w(TAG, "Quote source not found");
            quoteSource = "";
        } else if (TextUtils.isEmpty(sourceLeft)) {
            quoteSource = "―" + sourceRight;
        } else if (TextUtils.isEmpty(sourceRight)) {
            quoteSource = "―" + sourceLeft;
        } else {
            quoteSource = String.format("―%s, %s", sourceLeft, sourceRight);
        }

        return new QuoteData(quoteText, quoteSource);
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/modules/freakuotes/FreakuotesQuoteModule.kt
package com.crossbowffs.quotelock.modules.freakuotes

import android.content.ComponentName
import android.content.Context
import com.crossbowffs.quotelock.R
import com.crossbowffs.quotelock.api.QuoteData
import com.crossbowffs.quotelock.api.QuoteModule
import com.crossbowffs.quotelock.modules.freakuotes.FreakuotesQuoteModule
import com.crossbowffs.quotelock.utils.Xlog
import com.crossbowffs.quotelock.utils.downloadUrl
import org.jsoup.Jsoup

class FreakuotesQuoteModule : QuoteModule {

    companion object {
        private val TAG = FreakuotesQuoteModule::class.simpleName
    }

    override fun getDisplayName(context: Context): String? {
        return context.getString(R.string.module_freakuotes_name)
    }

    override fun getConfigActivity(context: Context): ComponentName? {
        return null
    }

    override fun getMinimumRefreshInterval(context: Context): Int {
        return 0
    }

    override fun requiresInternetConnectivity(context: Context): Boolean {
        return true
    }

    @Throws(Exception::class)
    override fun getQuote(context: Context): QuoteData? {
        val html = "https://freakuotes.com/frase/aleatoria".downloadUrl()
        val document = Jsoup.parse(html)
        val quoteContainer = document.select(".quote-container > blockquote").first()
        val quoteText = quoteContainer.getElementsByTag("p").text()
        if (quoteText.isNullOrEmpty()) {
            Xlog.e(TAG, "Failed to find quote text")
            return null
        }
        val sourceLeft = quoteContainer.select("footer > span").text()
        val sourceRight = quoteContainer.select("footer > cite").attr("title")
        val quoteSource: String = when {
            sourceLeft.isNullOrEmpty() && sourceRight.isNullOrEmpty() -> {
                Xlog.w(TAG, "Quote source not found")
                ""
            }
            sourceLeft.isNullOrEmpty() -> "― $sourceRight"
            sourceRight.isNullOrEmpty() -> "― $sourceLeft"
            else -> "― $sourceLeft, $sourceRight"
        }
        return QuoteData(quoteText, quoteSource)
    }

    override val characterType: Int
        get() = QuoteModule.CHARACTER_TYPE_LATIN
}